### DISCUSSION TITLE
"I'm a teapot" HTTP 418 when trying to add attachment to transaction in Firefly instance hosted on Dreamhost

### DISCUSSION THREAD

**Discussion by ChrisCarini** [firefly-iii/firefly-iii#4332] [Firefly III] [answered]
**Bug description**
I am running Firefly III version 5.3.3, and my problem is that I am getting "I'm a teapot" HTTP 418 when trying to add attachment to transaction.

**Steps to reproduce**
I edit a transaction, upload a pdf file, click update, and get the below banner, displayed for a *very* brief period of time, and then the page redirects saying the transaction was updated successfully (the pdf is not attached to the transaction, but any other updates are).

![Screen Shot 2021-01-31 at 15 24 47](https://user-images.githubusercontent.com/6374067/106401355-875c7000-63d8-11eb-8790-fcc9c1a00fcd.png)

**Extra info**
```
OS: OSX version 10.15.7
Chrome Version: Version 88.0.4324.96 (Official Build) (x86_64)
```
**`/debug` Output:**
Debug information generated at 2021-01-31 15:22:31 America/Los_Angeles for Firefly III version **5.3.3**.

| Scope | Version |
| --- | --- |
| Firefly III | 5.3.3 |
| Firefly III API | 1.3.0 |
| PHP | 7.4.14 |
| Host | Linux |

| System info | Value |
| --- | --- |
| Installation ID | <REDACTED - Let me know if this is required; I'd rather not post publicly.> |
| Using docker? | false |
| Telemetry | false |
| Layout |  |
| App environment | production |
| App debug mode | true |
| App cache driver | file |
| App logging | debug, daily |
| Display errors | Off |
| Error reporting | ALL errors |
| Interface | cgi-fcgi |
| Default language | en_US |
| Default locale | equal |
| BCscale | 12 |
| DB drivers | mysql, sqlite |
| Current driver | mysql |
| Login provider | eloquent  |
| Trusted proxies (.env) |  |

| User info | Value |
| --- | --- |
| Session start | 2021-01-01 00:00:00 |
| Session end | 2021-01-31 23:59:59 |
| Session first | 2018-12-01 00:00:00 |
| User ID | 1 |
| User language | en_US |
| User locale | en_US |
| Attempt at "en_US.utf8" | 'en_US.utf8' |
| Attempt at "en_US.UTF-8" | 'en_US.UTF-8' |
| User agent | Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/88.0.4324.96 Safari/537.36 |

**Bonus points**

- [x] I searched and nobody reported this bug before
- [ ] I have added a stack trace from my log files (***I think the log files may have been truncated too much, as nothing talks about an attachment***) 
- [x] I have added a screenshot.
- [ ] I was able to replicate it on the demo site https://demo.firefly-iii.org/  (***Demo site does not appear to allow attachments.***)

**Comment by JC5** [comment]
There's no 418 code sent out by Firefly III anywhere near that piece of code. There's one line in a piece of code controlling recurring transactions, but it's an "unreachable" line so it can't be sent out anyway.

I'm curious to see more logs on this.

**Reply by ChrisCarini** [reply]
@JC5 - can you advise how to gather logs for this? As mentioned, when I hit `/debug` there were no lines that seemed to indicate anything interesting/useful. Is there perhaps a more verbose output level or anything?

**Comment by JC5** [comment]
Also, please upgrade to the latest version.

**Reply by ChrisCarini** [reply]
@JC5 - I do plan to upgrade, however it's a non-trivial task for me (I have this setup on a shared hosting server of mine, and I've typically needed to compile a custom build of PHP to get firefly to install correctly). I likely won't have time to attempt this for at least a few weeks or so.

What's strange, is I was totally able to upload attachments on other transactions previously (on this same version/install).

**Comment by ChrisCarini** [comment]
Hi @JC5 - I have still been running into this issue. I just upgraded to the latest version (`5.6.14` as of writing) and encounter the same issue.

I just discovered the `/attachments` page, and noticed something quite interesting - it seems that the transaction IDs for all the 'failed to upload' attachments are incorrect! They are actually **ahead** of the most recent transaction.

Example, I just created `/transactions/show/5607` and attempted to upload a file 5 times - on the `/attachments` page, the link under the "Attached to" column links to `/transactions/show/5682` (which, does not exist - 75 above the most recent transaction). I went back and looked at the 3 other transactions where I tried to upload (ranging from 2020 to 2021, so not recent), and they too are all off-set (the other ones, though, are offset by 72).

I imagine there is something that is causing this in the app, but do not have any idea where to look in the code - could you help point in the right direction, or perhaps you know where this might be happening?

**Reply by ChrisCarini** [reply]
_Notes to self as I go about attempting to debug this (with very limited knowledge of the app, from a developers stand point); but hopefully this is useful to you as well, James!_

I see `Will now update journal #5682 (only journal in group #5607)` in the (debug) log, which the delta between these two numbers is 75 (`5682-5607 = 75`). As I look at older transactions with attachments, I also see a between the IDs - reading up on the 'concepts' (https://docs.firefly-iii.org/firefly-iii/concepts/transactions/) and 'architecture' (https://docs.firefly-iii.org/firefly-iii/about-firefly-iii/architecture/), I am realizing that these are likely different IDs (one for transaction, one for the journal - the difference has to do with transactions possibly being split (which, I do not do that often, hence why the delta between the 'two IDs' changing slowly over time).). So perhaps this is/was a red-herring.

## Local Changes
I modified `app/Api/V1/Controllers/Models/Attachment/StoreController.php` `store()` and `upload()` methods to try and get a bit more info; screenshot of modification below (simply printing vars):

<img width="1133" alt="Screen Shot 2022-02-13 at 17 45 11" src="https://user-images.githubusercontent.com/6374067/153786481-4955d535-d6a4-41f7-af79-d5ae4a55c243.png">

Then, going through the flow to upload an attachment again, I see the below when tailing the logs:
```
[2022-02-13 17:46:03] production.DEBUG: ================================DATA===========================  
[2022-02-13 17:46:03] production.DEBUG: array (
  'filename' => 'Screen Shot 2022-02-13 at 14.01.36.png',
  'title' => '',
  'notes' => '',
  'attachable_type' => 'TransactionJournal',
  'attachable_id' => 5682,
)  
[2022-02-13 17:46:03] production.DEBUG: ================================DATA===========================  
[2022-02-13 17:46:03] production.DEBUG: ================================attachment===========================  
[2022-02-13 17:46:03] production.DEBUG: {"user_id":1,"attachable_id":5682,"attachable_type":"FireflyIII\\Models\\TransactionJournal","md5":"","filename":"Screen Shot 2022-02-13 at 14.01.36.png","title":null,"description":null,"mime":"","size":0,"uploaded":false,"updated_at":"2022-02-14T01:46:03.000000Z","created_at":"2022-02-14T01:46:03.000000Z","id":46}  
[2022-02-13 17:46:03] production.DEBUG: ================================attachment===========================  
<END_OF_TAIL_LOG>
```

This would lead me to believe that something is going on before the app enters the `upload()` method - based on what I saw in `api.php` (specifically, the line: `Route::post('{attachment}/upload', ['uses' => 'StoreController@upload', 'as' => 'upload']);`), I believe this is the correct method that should be executing.

I'm not that familiar w/ the app and it's supporting frameworks/infra, so I'm not sure where to look next. Any guidance would be greatly appreciated!! :)

**Reply by JC5** [reply]
Thanks for the extensive research!

I knew this was something that happens before the transaction reaches Firefly III. Now I know for sure.

> I'm not that familiar w/ the app and it's supporting frameworks/infra, so I'm not sure where to look next. Any guidance would be greatly appreciated!! :)

Let's start by telling me about your infra :+1:

**Reply by ChrisCarini** [reply]
> Let's start by telling me about your infra 👍

SGTM!


### Firefly-iii & `php` versions, along with required php packages

I am running firefly using the self-hosted option, currently running version `5.6.14` on php version `8.0.2` which includes all of the needed supported packages (excerpt of `php -i` below):
```
$ /usr/local/php80/bin/php -i | grep -i "version\|bcmath\|internationalization\|curl\|zip\|gd\|xml\|mbstring\|ldap\|mysql"| grep -i "support => "
Zend Multibyte Support => provided by mbstring
BCMath support => enabled
BZip2 Support => Enabled
Stream Wrapper support => compress.bzip2://
Stream Filter support => bzip2.decompress, bzip2.compress
cURL support => enabled
GD Support => enabled
Internationalization support => enabled
libXML support => active
MysqlI Support => enabled
SimpleXML support => enabled
XML Support => active
XML Namespace Support => active
```

Additionally, I am running this on a shared hosting plan which uses php 8 + fastcgi.

### `firefly-iii/storage` permissions
Per the instructions, the `firefly-iii/storage` directory was configured with the correct permissions:
```shell
$ ls -lash firefly-iii/storage
total 24K
4.0K drwxrwxr-x 10 <expected_user> <expected_group> 4.0K Feb 13 10:27 .
4.0K drwxrwxr-x 13 <expected_user> <expected_group> 4.0K Feb 13 10:25 ..
   0 drwxrwxr-x  3 <expected_user> <expected_group>   48 Feb  5 00:10 app
   0 drwxrwxr-x  2 <expected_user> <expected_group>   78 Feb  5 00:10 build
   0 drwxrwxr-x  2 <expected_user> <expected_group>   31 Feb  5 00:10 database
   0 drwxrwxr-x  2 <expected_user> <expected_group>   31 Feb  5 00:10 debugbar
   0 drwxrwxr-x  2 <expected_user> <expected_group>   31 Feb  5 00:10 export
   0 drwxrwxr-x  6 <expected_user> <expected_group>  100 Feb  5 00:10 framework
4.0K drwxrwxr-x  2 <expected_user> <expected_group> 4.0K Feb 14 00:03 logs
4.0K -rwxrwxr-x  1 <expected_user> <expected_group> 3.3K Feb 13 10:27 oauth-private.key
4.0K -rwxrwxr-x  1 <expected_user> <expected_group>  812 Feb 13 10:27 oauth-public.key
4.0K drwxrwxr-x  2 <expected_user> <expected_group> 4.0K Feb 13 10:27 upload
```

### Latest info from `/debug` endpoint
Debug information generated at 2022-02-14 08:24:28 America/Los_Angeles for Firefly III version **5.6.14**.

| Scope | Version |
| --- | --- |
| Firefly III | 5.6.14 |
| Firefly III API | 1.5.5 |
| Build | (unknown), (unknown) |
| PHP | 8.0.2 |
| Host | Linux |

| System info | Value |
| --- | --- |
| Installation ID | 8716776e-67f3-471e-9c07-53782b81bdd1 |
| Using docker? | false |
| Layout |  |
| System TZ | America/Los_Angeles |
| Browser TZ | America/Los_Angeles |
| App environment | production |
| App debug mode | true |
| App cache driver | file |
| App logging | debug, daily |
| Display errors | Off |
| Error reporting | ALL errors |
| Interface | cgi-fcgi |
| Default language | en_US |
| Default locale | equal |
| BCscale | 24 |
| DB drivers | mysql, sqlite |
| Current driver | mysql |
| DB version | 18 (exp. 18) |
| Login provider | eloquent  |
| Trusted proxies (.env) |  |

| User info | Value |
| --- | --- |
| Session start | 2022-02-01 00:00:00 |
| Session end | 2022-02-28 23:59:59 |
| Session first | 2017-11-01 00:00:00 |
| User ID | 1 |
| User language | en_US |
| User locale | en_US |
| Attempt at "en_US.utf8" | 'en_US.utf8' |
| Attempt at "en_US.UTF-8" | 'en_US.UTF-8' |
| User agent | Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/98.0.4758.80 Safari/537.36 |


@JC5 - Please let me know if there is any additional information you need from me! I am happy to help!

**Reply by JC5** [reply]
https://help.dreamhost.com/hc/en-us/articles/215947927-How-do-I-enable-Extra-Web-Security-for-my-website-

**Reply by ChrisCarini** [reply]
🤦  _That's embarrassing._

![Screen Shot 2022-02-14 at 17 48 15](https://user-images.githubusercontent.com/6374067/153977339-d09322d9-6f91-443e-92b7-5184a8b4cfd6.png)
![Screen Shot 2022-02-14 at 17 48 31](https://user-images.githubusercontent.com/6374067/153977344-f26ee1af-b81c-41f2-857b-8094fa82bf8e.png)

**_Totally works now!_** 🎉 

Thank you @JC5 ! I will mark this as answered, and update the original question/title to include `Dreamhost` since this is specific to the hosting provider I am using on this installation. Very much appreciate you helping and finding that! Totally out of scope of the project bounds, IMO, so I very much appreciate it!

**Reply by JC5** [reply]
No problem, glad to hear it works now! It took your hints to find the answer after all, so not all credits are for me.

**Comment by ChrisCarini** [comment, marked as answer]
_🎉 **Credit to @JC5 for finding the answer!** ([credit where credit is due 🪙; see comment in the above thread here](https://github.com/firefly-iii/firefly-iii/discussions/4332#discussioncomment-2175065)) 🎉_

**TL;DR:** When using `Dreamhost` as your hosting provider, you need to disable `Extra Web Security` in your sites configuration. [See this Dreamhost help article for details](https://help.dreamhost.com/hc/en-us/articles/215947927-How-do-I-enable-Extra-Web-Security-for-my-website-).
